File: project_materials.py
"""Суммарный расчёт количества материалов по проекту (как на экране полного расчёта)."""
from models import CeilingLayout
from materials_calculator import (
    calculate_materials,
    room_perimeter_cm,
    room_area_m2,
    grilyato_cassette_count,
)

TYPE_MAP = {"Армстронг": "armstrong", "Грильято": "grilyato_classic", "GL": "grilyato_gl"}


def room_effective_config(room, project):
    p_ceiling = getattr(project, "materials_ceiling", None) or "Армстронг"
    p_susp = getattr(project, "materials_susp", None) or "Подвес 0,5"
    p_cell = getattr(project, "materials_cell", None) or "50x50"
    if bool(getattr(room, "materials_override", False)):
        return (
            getattr(room, "materials_ceiling", None) or p_ceiling,
            getattr(room, "materials_susp", None) or p_susp,
            getattr(room, "materials_cell", None) or p_cell,
        )
    return p_ceiling, p_susp, p_cell


def normalize_result_key(name, value, ceiling, susp, cell):
    key = name
    if name == "Подвес":
        size_txt = (susp or "Подвес 0,5").replace("Подвес", "").strip()
        key = f"Подвес ({size_txt})"
    elif name in ("Профиль Папа", "Профиль Мама", "Заглушки"):
        key = f"{name} ({cell})"
    elif name == "Плиты/кассеты":
        key = f"{name} ({ceiling})"
    elif name == "Светильники":
        key = "Светильники"
    return key, int(value)


def aggregate_project_totals(project):
    """
    Возвращает (totals: dict[str,int], area_m2, perimeter_m).
  totals — ключи как в экране полного расчёта.
    """
    totals = {}
    total_area = 0.0
    total_perimeter_m = 0.0

    if not project or not getattr(project, "rooms", None):
        return totals, total_area, total_perimeter_m

    for room in project.rooms:
        if not getattr(room, "walls", None):
            continue

        ceiling, susp, cell = room_effective_config(room, project)
        calc_type = TYPE_MAP.get(ceiling, "armstrong")

        layout = CeilingLayout(room)
        layout.calculate_layout()
        cassette_count = layout.full_tiles + layout.cut_tiles
        if calc_type in ("grilyato_gl", "grilyato_classic"):
            cassette_count = grilyato_cassette_count(room.walls, cassette_count)

        light_count = len(getattr(room, "light_fixtures", []) or [])

        result = calculate_materials(
            ceiling_type=calc_type,
            walls=room.walls,
            cassette_count=cassette_count,
            rows_3600=None,
            rows_2400=None,
            cell_size=cell,
            light_count=light_count,
        )

        total_area += room_area_m2(room.walls)
        total_perimeter_m += room_perimeter_cm(room.walls) / 100.0

        for name, value in result.items():
            key, qty = normalize_result_key(name, value, ceiling, susp, cell)
            totals[key] = totals.get(key, 0) + qty

    return totals, total_area, total_perimeter_m

File: flutter_app/lib/domain/project_materials.dart
import 'ceiling_grid.dart';
import 'ceiling_guides.dart';
import 'materials_calculator.dart';
import 'models.dart';

const typeMap = {
  'Армстронг': CeilingType.armstrong,
  'Грильято': CeilingType.grilyatoClassic,
  'GL': CeilingType.grilyatoGl,
};

(String ceiling, String susp, String cell) roomEffectiveConfig(Room room, Project project) {
  final pCeiling = project.materialsCeiling ?? 'Армстронг';
  final pSusp = project.materialsSusp ?? 'Подвес 0,5';
  final pCell = project.materialsCell ?? '50x50';
  if (room.materialsOverride) {
    return (
      room.materialsCeiling ?? pCeiling,
      room.materialsSusp ?? pSusp,
      room.materialsCell ?? pCell,
    );
  }
  return (pCeiling, pSusp, pCell);
}

(String key, int qty) normalizeResultKey(
  String name,
  int value,
  String ceiling,
  String susp,
  String cell,
) {
  var key = name;
  if (name == 'Подвес') {
    final sizeTxt = susp.replaceFirst('Подвес', '').trim();
    key = 'Подвес ($sizeTxt)';
  } else if (name == 'Профиль Папа' || name == 'Профиль Мама' || name == 'Заглушки') {
    key = '$name ($cell)';
  } else if (name == 'Плиты/кассеты') {
    key = '$name ($ceiling)';
  } else if (name == 'Светильники') {
    key = 'Светильники';
  }
  return (key, value);
}

List<(double, double)> pointsFromWalls(List<List<double>> walls) {
  return MaterialsCalculator.orderedPointsFromWalls(walls);
}

/// Layout-driven materials for one room.
Map<String, int> calculateRoomMaterials(Room room, Project project) {
  if (room.walls.length < 3) return {};
  final (ceiling, susp, cell) = roomEffectiveConfig(room, project);
  final type = typeMap[ceiling] ?? CeilingType.armstrong;

  final grid = CeilingGrid(
    offsetXCm: room.gridOffsetX.toDouble(),
    offsetYCm: room.gridOffsetY.toDouble(),
  );
  final lights = <String>{};
  for (final f in room.lightFixtures) {
    if (f.length >= 2) lights.add('${f[0]}:${f[1]}');
  }
  grid.rebuild(
    roomPoints: pointsFromWalls(room.walls),
    walls: room.walls,
    lightIds: lights,
  );
  grid.pruneLights();

  final layoutTiles = grid.cells
      .where((c) => c.kind == CellKind.full || c.kind == CellKind.cut)
      .map(
        (c) => (
          isFull: c.kind == CellKind.full,
          cutXCm: c.cutXCm,
          cutYCm: c.cutYCm,
        ),
      );
  final cassette = MaterialsCalculator.estimatePurchaseCassettes(
    walls: room.walls,
    layoutTiles: layoutTiles,
    cellCm: grid.cellSizeCm,
  );
  final lightCount = room.lightFixtures.length;
  final raw = MaterialsCalculator.calculate(
    type: type,
    walls: room.walls,
    cassetteCount: cassette,
    cellSize: cell,
    lightCount: lightCount,
  );

  final guideMarks = parseCeilingGuidesWithLegacyAnchor(
    room.ceilingGuides,
    room.walls,
    offsetX: room.gridOffsetX.toDouble(),
    offsetY: room.gridOffsetY.toDouble(),
    cellSize: grid.cellSizeCm,
  );
  final guideCounts = countGuidesFromMarks(
    guideMarks,
    room.walls,
    offsetX: room.gridOffsetX.toDouble(),
    offsetY: room.gridOffsetY.toDouble(),
    cellSize: grid.cellSizeCm,
  );
  final adjusted = applyGuideOverrides(raw, guideCounts, type: type);

  final out = <String, int>{};
  for (final e in adjusted.entries) {
    final (key, qty) = normalizeResultKey(e.key, e.value, ceiling, susp, cell);
    out[key] = (out[key] ?? 0) + qty;
  }
  return out;
}

(Map<String, int> totals, double areaM2, double perimeterM) aggregateProjectTotals(Project project) {
  final totals = <String, int>{};
  var area = 0.0;
  var peri = 0.0;
  for (final room in project.rooms) {
    if (room.walls.length < 3) continue;
    area += MaterialsCalculator.roomAreaM2(room.walls);
    peri += MaterialsCalculator.roomPerimeterCm(room.walls) / 100.0;
    final part = calculateRoomMaterials(room, project);
    for (final e in part.entries) {
      totals[e.key] = (totals[e.key] ?? 0) + e.value;
    }
  }
  return (totals, area, peri);
}
